src/jltypes.c

```diff
@@ -1287,7 +1306,10 @@ static jl_value_t *inst_datatype_inner(jl_datatype_t *dt, jl_svec_t *p, jl_value
         }
         else if (cacheable) {
             // recursively instantiate the types of the fields
-            ndt->types = inst_ftypes(ftypes, env, stack);
+            if (dt->types == NULL)
+                ndt->types = jl_compute_fieldtypes(ndt);
```

Comment:
Curious why this is needed?

Reply:
It's reveled from an interaction of the PR with lazy-ftypes (although just the type-stack alone could theoretically run into the same problem with this PR). Previously, on arriving in this function, the dt was always the wrapper object and so the env was already that which will be created by `jl_compute_fieldtypes`. With the PR, it now could be a partially computed type that just needs to be copied, but that also means it could now be the final parameter application to create a concrete type (where with lazy-ftypes, we are now not been computing the ftypes of that intermediate work).